Firestore/core/src/firebase/firestore/immutable/sorted_map.h

```diff
@@ -0,0 +1,209 @@
+   */
+  explicit SortedMap(const C& comparator = C())
+      : SortedMap(array_type{comparator}) {
+  }
+
+  /**
+   * Creates an SortedMap containing the given entries.
+   */
+  SortedMap(std::initializer_list<value_type> entries,
+            const C& comparator = C()) {
+    if (entries.size() <= kFixedSize) {
+      tag_ = Tag::Array;
+      new (&array_) array_type(entries, comparator);
+    } else {
+      abort();
+    }
+  }
+
+  SortedMap(const SortedMap& other) : tag_(other.tag_) {
+    switch (tag_) {
+      case Tag::Array:
+        new (&array_) array_type(other.array_);
+        break;
+      case Tag::Tree:
+        new (&tree_) tree_type(other.tree_);
+        break;
+    }
+  }
+
+  SortedMap(SortedMap&& other) : tag_(other.tag_) {
+    switch (tag_) {
+      case Tag::Array:
+        new (&array_) array_type(std::move(other.array_));
+        break;
+      case Tag::Tree:
+        new (&tree_) tree_type(std::move(other.tree_));
+        break;
+    }
+  }
+
+  ~SortedMap() {
+    switch (tag_) {
+      case Tag::Array:
+        array_.~ArraySortedMap();
+        break;
+      case Tag::Tree:
+        tree_.~TreeSortedMap();
+        break;
+    }
+  }
+
+  SortedMap& operator=(const SortedMap& other) {
+    if (tag_ == other.tag_) {
+      switch (tag_) {
+        case Tag::Array:
+          array_ = other.array_;
+          break;
+        case Tag::Tree:
+          tree_ = other.tree_;
+          break;
+      }
+    } else {
+      this->~SortedMap();
+      new (this) SortedMap{other};
+    }
+    return *this;
+  }
+
+  SortedMap& operator=(SortedMap&& other) {
+    if (tag_ == other.tag_) {
+      switch (tag_) {
+        case Tag::Array:
+          array_ = std::move(other.array_);
+          break;
+        case Tag::Tree:
+          tree_ = std::move(other.tree_);
+          break;
+      }
+    } else {
+      this->~SortedMap();
+      new (this) SortedMap{std::move(other)};
+    }
+    return *this;
+  }
+
+  /**
+   * Creates a new map identical to this one, but with a key-value pair added or
+   * updated.
+   *
+   * @param key The key to insert/update.
+   * @param value The value to associate with the key.
+   * @return A new dictionary with the added/updated value.
+   */
+  SortedMap insert(const K& key, const V& value) const {
+    switch (tag_) {
+      case Tag::Array:
+        // TODO(wilhuff): convert to TreeSortedMap
+        return SortedMap{array_.insert(key, value)};
+      case Tag::Tree:
+        return SortedMap{tree_.insert(key, value)};
+    }
+  }
+
+  /**
+   * Creates a new map identical to this one, but with a key removed from it.
+   *
+   * @param key The key to remove.
+   * @return A new map without that value.
+   */
+  SortedMap erase(const K& key) const {
+    switch (tag_) {
+      case Tag::Array:
+        return SortedMap{array_.erase(key)};
+      case Tag::Tree:
+        return SortedMap{tree_.erase(key)};
+    }
+  }
+
+  /** Returns true if the map contains no elements. */
```

Comment:
Very optional: I know my opinion is peculiar on this, and feel free to ignore, but I'd remove this comment and the comment on `size` as well. Both of these functions do the intuitive thing, and also both follow the common STL convention.